### PR TITLE
Removing expired logs from mongo data base.

### PR DESCRIPTION
**Pull request by tatianaIljina**

### CODE REVIEW

File: API/Tenders.API/Controllers/LogsController.cs [hunk 1]
```diff
@@ -99,10 +100,14 @@ public IActionResult Post(ILogModel log)
                 return this.BadRequest("Sender cannot be empty");
             }
 
-            if (!this.repository.Create(log))
+            await Task.Factory.StartNew(async () =>
```
Comment (Anny777) [inline API/Tenders.API/Controllers/LogsController.cs]: Not sure if this wrapper is needed

File: API/Tenders.API/Services/CheckLastRespService.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public class CheckLastRespService : ICheckLastRespService
+    {
+        private DateTime LastRemoving { get; set; }
+
+        /// <inheritdoc/>
+        public bool IsNeedToBeClear()
+        {
+            var currentTime = DateTime.Now;
+            var launchTime = currentTime.Date.AddHours(22);
+            if (currentTime.Date <= this.LastRemoving.Date && currentTime < launchTime)
```
Comment (Anny777) [inline API/Tenders.API/Services/CheckLastRespService.cs]: Wrong comparison. It is better to swap and use ||

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 3]
```diff
@@ -65,13 +71,50 @@ public bool CleanLogs()
         }
 
         /// <inheritdoc/>
-        public bool Create(ILogModel log)
+        public async Task<bool> DeleteLogsAsync()
+        {
+            var builder = Builders<LogModel>.Filter;
+            var filter = builder.Where(r =>
+                (r.LogLevel == LogLevel.Fatal || r.LogLevel == LogLevel.Error) &&
```
Comment (Anny777) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: You need to check each level of the log.

File: API/Tenders.API/Controllers/LogsController.cs [hunk 4]
```diff
@@ -144,5 +146,22 @@ public IActionResult CleanLogs()
...
+
+        /// <summary>
+        /// Removes expired logs.
+        /// </summary>
+        /// <returns>Ok.</returns>
+        [HttpDelete]
+        [Route("RemoveExpirated")]
+        public async Task<IActionResult> RemoveExpirated()
+        {
+            bool res = await this.repository.DeleteLogsAsync();
```
Comment (Anny777) [inline API/Tenders.API/Controllers/LogsController.cs]: Use var

File: API/Tenders.API/Controllers/LogsController.cs [hunk 5]
```diff
@@ -144,5 +146,22 @@ public IActionResult CleanLogs()
...
+            if (res)
+            {
+                return this.Ok(true);
```
Comment (Anny777) [inline API/Tenders.API/Controllers/LogsController.cs]: Maybe you should not pass true. Return ok means the result is true.

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 6]
```diff
@@ -65,13 +71,50 @@ public bool CleanLogs()
...
+            await this.DeleteManyAsync(filter);
+
+            filter = builder.Where(r =>
+                r.LogLevel == LogLevel.Debug &&
+                r.LogDate <= DateTime.Now.AddDays(-14));
+
+            await this.DeleteManyAsync(filter);
+
+            filter = builder.Where(r =>
+                (r.LogLevel == LogLevel.Trace || r.LogLevel == LogLevel.Info) &&
```
Comment (Anny777) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: Fix storage for 3 days.

File: API/Tenders.API/Controllers/LogsController.cs [hunk 7]
```diff
@@ -144,5 +146,22 @@ public IActionResult CleanLogs()
...
             return this.Ok();
         }
+
+        /// <summary>
+        /// Removes expired logs.
+        /// </summary>
+        /// <returns>Ok.</returns>
+        [HttpDelete]
+        [Route("RemoveExpirated")]
+        public async Task<IActionResult> RemoveExpirated()
```
Comment (DamienVStone) [inline API/Tenders.API/Controllers/LogsController.cs]: Cleaning should run on Log act.

File: API/Tenders.API/DAL/Interfaces/IMongoLogsRepository.cs [hunk 8]
```diff
@@ -44,5 +45,11 @@ public interface IMongoLogsRepository : IRepository<LogModel>
         /// </summary>
         /// <returns>Result of clean logs.</returns>
         bool CleanLogs();
+
+        /// <summary>
+        /// Removes expered logs.
+        /// </summary>
+        /// <returns>The result of async operation.</returns>
+        Task<bool> DeleteLogsAsync();
```
Comment (DamienVStone) [inline API/Tenders.API/DAL/Interfaces/IMongoLogsRepository.cs]: Why you made `DeleteLogsAsync()` if interface already contains `CleanLogs()`?

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 9]
```diff
@@ -202,5 +233,14 @@ private IEnumerable<LogModel> GetData()
             var res = this.Entities.FindSync(filter, null, default(CancellationToken)).ToList();
             return res;
         }
+
+        private async Task<bool> DeleteManyAsync(FilterDefinition<LogModel> filter)
+        {
+            return await Task.Run<bool>(() =>
```
Comment (DamienVStone) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: What is it?

File: API/Tenders.API/Services/CheckLastRespService.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Tenders.API.Services
+{
+    using System;
+    using Tenders.API.Services.Interfaces;
+
+    /// <inheritdoc/>
+    public class CheckLastRespService : ICheckLastRespService
+    {
+        private DateTime LastRemoving { get; set; }
```
Comment (DamienVStone) [inline API/Tenders.API/Services/CheckLastRespService.cs]: If your service is singleton and keep some state (as LastRemoving) you should guarantee thread safety.

File: tests/Tenders.API.Tests/CheckLastRespServiceTest.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            Assert.True(res);
+        }
+
+        /// <summary>
+        /// Determines the necessarity when response did this day.
+        /// </summary>
+        [Fact]
+        public void DeterminesStatusFalse()
+        {
+            typeof(CheckLastRespService).GetProperty("LastRemoving", BindingFlags.NonPublic | BindingFlags.Instance)
```
Comment (DamienVStone) [inline tests/Tenders.API.Tests/CheckLastRespServiceTest.cs]: Lol, what if you just call `IsNeedToBeClear()` twice? =)

File: tests/Tenders.API.Tests/MongoLogsRepositoryTest.cs [hunk 12]
```diff
@@ -247,6 +261,23 @@ public void Clean()
...
         }
 
+        /// <summary>
+        /// Can remove logs.
+        /// </summary>
+        /// <returns>
+        /// Result of async operation.
+        /// </returns>
+        [Fact]
+        public async Task CanRemoveExperedLogs()
```
Comment (DamienVStone) [inline tests/Tenders.API.Tests/MongoLogsRepositoryTest.cs]: What means CanRemove**Expered**Logs?

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 13]
```diff
@@ -42,36 +45,37 @@ public class MongoLogsRepository : BaseMongoRepository<LogModel>, IMongoLogsRepo
...
-        /// <inheritdoc/>
-        public bool Create(ILogModel log)
+        public async Task<bool> Create(ILogModel log)
         {
             var model = this.factory.MakeInstance() as LogModel;
             model.Sender = log.Sender;
             model.LogMessage = log.LogMessage;
             model.LogLevel = log.LogLevel;
             model.LogDate = log.LogDate;
+            var isClear = this.checkLastRespService.IsNeedToBeClear();
```
Comment (DamienVStone) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: "is clear" equals "is need to be clear"?

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 14]
```diff
@@ -42,36 +45,37 @@ public class MongoLogsRepository : BaseMongoRepository<LogModel>, IMongoLogsRepo
...
+            if (isClear)
```
Comment (DamienVStone) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: if is clear then clear?

File: API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs [hunk 15]
```diff
@@ -42,36 +45,37 @@ public class MongoLogsRepository : BaseMongoRepository<LogModel>, IMongoLogsRepo
...
+            {
+                var count = await this.ClearExpiredLogsAsync();
+                await this.logger.Log($"Was deleted {count} logs.");
```
Comment (DamienVStone) [inline API/Tenders.API/DAL/Mongo/MongoLogsRepository.cs]: Change to Info

**Pull request merged**